File: Java/Leetcode994.java
package Java;

import java.util.LinkedList;
import java.util.Queue;

public class Leetcode994 {
    public int orangesRotting(int[][] grid){
        if(grid==null || grid.length == 0 || grid[0].length == 0) {return 0;}

        int[] dx = {0, 1, 0, -1};
        int[] dy = {1, 0, -1, 0};
        int row = grid.length, col = grid[0].length;

        Queue<int[]> queue = new LinkedList<>();

        // get rotten orange's coordinate
        for(int i=0; i<row; i++) {
            for (int j = 0; j < col; j++) {
                if (grid[i][j] == 2) {
                    queue.offer(new int[]{i, j});
                }
            }
        }

        int count = 0;

        while(!queue.isEmpty()){
            // 한 번 전염되는 만큼만 꺼낼 것
            // Do not get every element in queue
            int size = queue.size();
            for(int s=0; s<size; s++){
                int[] curr = queue.poll();

                for(int d=0; d<4; d++){
                    int nr = curr[0] + dx[d];
                    int nc = curr[1] + dy[d];

                    if(isValid(grid, nr, nc)){
                        grid[nr][nc] = 2;
                        queue.add(new int[]{nr, nc});
                    }
                }
            }
            count += 1;
        }

        // If fresh orange remaining
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(grid[i][j] == 1) {return -1;}
            }
        }

        // count-1 because we added 1 in last recursion
        return count==0 ? 0 : count - 1;

    }

    public boolean isValid(int[][] grid, int i, int j){
        return i>=0 && i<grid.length && j>=0 && j<grid[0].length && grid[i][j] == 1;
    }
}

File: Python/Leetcode994.py
class Leetcode994(object):
    def orangesRotting(self, grid):
        m, n = len(grid), len(grid[0])
        count = 0
        q = collections.deque()

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    # Number of left fresh orange
                    count += 1
                elif grid[i][j] == 2:
                    q.append((i,j,0))
        
        visited = set()

        # (x, y, d): x, y는 좌표, d는 해당 좌표가 썩은 시간
        while q:
            x, y, d = q.popleft()
            for x, y in ((x+1, y), (x-1, y), (x, y+1), (x, y-1)):
                if 0<=x<m and 0<=y<n and (x,y) not in visited and grid[x][y] == 1:
                    visited.add((x, y))
                    count -= 1
                    if count == 0:
                        return d+1
                    q.append((x, y, d+1))
        # if no fresh orange -> return 0, else if fresh orange left -> -1
        return 0 if count == 0 else -1
